Sources/TABResourceLoader/Protocols/Resource types/NetworkJSONResourceType.swift

```diff
@@ -0,0 +1,51 @@
+//
+//  NetworkJSONResourceType.swift
+//  TABResourceLoader
+//
+//  Created by John Sanderson on 30/01/2018.
+//
+//  The MIT License (MIT)
+//
+//  Copyright (c) 2016 Luciano Marisi
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included
+//  in all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+
+import Foundation
+
+/// Defines a resource that can be used to POST JSON bodt to a network
+public protocol NetworkJSONResourceType: NetworkResourceType {
+
+  /// The JSON body used to fetch this resource
+  var jsonBody: Any? { get }
+}
+
+public extension NetworkJSONResourceType {
+  var jsonBody: Any? { return nil }
+
+  var httpHeaderFields: [String: String]? {
+    return ["Content-Type": "application/json"]
```

Comment:
i've never actually liked the fact that we had this in there in the first place. i'm sure most users of this library are sending GET requests without sending content.

it would be better to distinguish between the request content type and the response content type.

e.g. would you ever need to send FORM data and get a JSON response? perhaps. the current implementation restricts this by confusing the two. this isn't anything to do with your change, just an observation about the restriction of the library in the first place. we've definitely gone with the simplest approach, which serve 90% of use cases. but something to bear in mind.

Reply:
on second thoughts, i think it would be good to give this more thought now.

looking at your `NetworkFormDataResourceType`, it conforms to `NetworkResourceType` and `DataResourceType`. i assume you just use this to send form data and get an HTTP response code without any content in the response body?

what if you did get some content in the response body? what about error messages? do you get any JSON to give you more details on the error?

i think it would be good to think about this in terms of `DataRequestType` and `DataResponseType` protocols, which we could extend with `FormDataRequestType`, `JSONDataRequestType`, `FormDataResponseType` and `FormDataResponseType`. do you ever get form data in a response?

the "Request" type protocols would implement in protocol extensions the `var body: Data? { get }` requirement of the `DataRequestType` protocol, just as your `NetworkJSONResourceType` extension does here. these extensions would also implement the relevant header fields' content type.

the "Response" type protocols would implement in protocol extensions the `func model(from data: Data) throws -> Model` requirement of the `DataResponseType` protocol, just as the `JSONDictionaryResourceType` currently does.

Reply:
@samdods did you read my comment above? looks like we're on a similar page...

Reply:
yeah just read your comment after writing mine 👍